### PR TITLE
Implement path-bases (RFC 3529) 2/n: support for nested packages

### PR DESCRIPTION
**Pull request by dpaoliello**
Cargo has the ability to discover "nested packages" when loading the dependencies of a package. For example, if a git dependency has a path dependency pointing to somewhere else in the repo, or a submodule, then Cargo needs to discover those dependencies as well.

This change makes the "nested packages" loading code aware of the "path bases" feature so that it can build the correct manifest path for any path dependencies that may be using a `base`.

RFC: https://github.com/rust-lang/rfcs/pull/3529
Tracking Issue: https://github.com/rust-lang/cargo/issues/14355

### CODE REVIEW

**Comment by epage** [comment]
> Cargo has the ability to discover "nested packages" when loading the dependencies of a package. For example, if a git dependency has a path dependency pointing to somewhere else in the repo, or a submodule, then Cargo needs to discover those dependencies as well.
>
> This change makes the "nested packages" loading code aware of the "path bases" feature so that it can build the correct manifest path for any path dependencies that may be using a base.
>
> Note that this change also fixes a bug in the "nested package" loading code: specifically, that all nested packages were read using the GlobalContext of the "root" package, effectively ignoring any config.toml files of those nested packages.

I believe the existing behavior is what is "expected" and this PR is a regression in Cargo.

**Comment by dpaoliello** [comment]
> I believe the existing behavior is what is "expected" and this PR is a regression in Cargo.
> 
> EDIT: To expand on this, configuration is environmental configuration and not project configuration. The latter is what `Cargo.toml` is for but we aren't supporting `Cargo.toml` yet. See also #14001, #2930

Ah, ok, #2930 is what I was tripping over when writing my test: I'll adjust the test and revert the regression.

I'll also add a note to the docs to mention this: if you depend on a package that uses a path base, then that path base must be defined in the configuration loaded for the root package.

File: tests/testsuite/git.rs [hunk 1]
```diff
@@ -380,6 +382,109 @@ hello world
...
+    git::add_submodule(&repo, &url, Path::new("submods/dep2"));
+    git::commit(&repo);
+
+    write_config_at(
+        paths::home().join(".cargo/config.toml"),
+        r#"
+            [path-bases]
+            submodules = '../dep1/submods'
+        "#,
+    );
```
Comment (epage) [inline tests/testsuite/git.rs:446]: imo writing path-bases on behalf of git dependencies seems very questionable
- Constructing the right absolute path is a mess
- Constructing the right relative path is a mess

If you want to test `RecursivePathSource`, https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#paths-overrides is another user of that tyoe

File: src/cargo/sources/path.rs [hunk 2]
```diff
@@ -1000,8 +1005,31 @@ fn read_nested_packages(
...
+                    Ok(base) => Cow::Owned(base.join(p)),
+                    Err(err) => {
+                        errors.push(err);
+                        continue;
+                    }
+                }
+            } else {
+                Cow::Borrowed(p)
+            };
+            let path = paths::normalize_path(&path.join(p.as_path()));
```
Comment (epage) [inline src/cargo/sources/path.rs:1032]: Why is this needed?  Aren't we operating on a normalized package and don't we resolve path-bases during normaization?

For either of us to verify that is one of the reasons I recommend splitting commits into
- Test showing the existing behavior
- Fix/feature with test updates to show how the behavior changed

**Comment by dpaoliello** [comment]
Looks like this change isn't needed: #2930 really confused me.
Sorry for the noise!
